src/rust/ide/src/controller/visualization.rs

```diff
@@ -0,0 +1,254 @@
+//! `Visualization Controller` lives as long as the `Project Controller`.
+
+use crate::prelude::*;
+
+use std::rc::Rc;
+use enso_protocol::language_server;
+use graph_editor::component::visualization::class;
+use graph_editor::component::visualization::JsSourceClass;
+
+
+
+// =============
+// === Error ===
+// =============
+
+/// Enumeration of errors used in `Visualization Controller`.
+#[derive(Debug,Fail)]
+#[allow(missing_docs)]
+pub enum VisualizationError {
+    #[fail(display = "Visualization \"{}\" not found.", identifier)]
+    NotFound {
+        identifier : VisualizationIdentifier
+    },
+    #[fail(display = "JavaScript visualization \"{}\" failed to compile.", identifier)]
+    CompileError {
+        identifier : VisualizationIdentifier
+    }
+}
+
+
+
+// ===============================
+// === VisualizationIdentifier ===
+// ===============================
+
+/// This enum is used to identify a visualization both in the project folder or natively embedded in
+/// IDE.
+#[derive(Clone,Debug,Display,Eq,PartialEq)]
+#[allow(missing_docs)]
+pub enum VisualizationIdentifier {
+    Embedded(String),
+    File(language_server::Path)
+}
+
+
+
+// ==============================
+// === EmbeddedVisualizations ===
+// ==============================
+
+#[allow(missing_docs)]
+pub type EmbeddedVisualizationName   = String;
+
+/// Embedded visualizations mapped from name to source code.
+#[derive(Shrinkwrap,Debug,Clone,Default)]
+#[shrinkwrap(mutable)]
+pub struct EmbeddedVisualizations {
+    #[allow(missing_docs)]
+    pub map : HashMap<EmbeddedVisualizationName,Rc<class::Handle>>
+}
+
+
+
+
+// ==============
+// === Handle ===
+// ==============
+
+const VISUALISATION_FOLDER : &str = "visualization";
+
+#[derive(Debug)]
+struct VisualizationControllerData {
+    language_server_rpc     : Rc<language_server::Connection>,
+    embedded_visualizations : EmbeddedVisualizations
+}
+
+/// Visualization Controller's state.
+#[derive(Debug,Clone,CloneRef)]
+pub struct Handle {
+    data : Rc<RefCell<VisualizationControllerData>>
+}
+
+impl Handle {
+    /// Creates a new visualization controller.
+    pub fn new
+    ( language_server_rpc     : Rc<language_server::Connection>
+    , embedded_visualizations : EmbeddedVisualizations) -> Self {
+        let data = VisualizationControllerData{language_server_rpc,embedded_visualizations};
+        let data = Rc::new(RefCell::new(data));
+        Self {data}
+    }
+
+    async fn list_file_visualizations(&self) -> FallibleResult<Vec<VisualizationIdentifier>> {
+        let data                = self.data.borrow();
+        let language_server_rpc = &data.language_server_rpc;
+        let root_id             = language_server_rpc.content_root();
+        let path                = language_server::Path::new(root_id,&[VISUALISATION_FOLDER]);
+        let file_list           = language_server_rpc.file_list(&path).await?;
```

Comment:
I'm not convinced if we should error-out if the visualization folder does not exist. Perhaps we should just return empty vector then.
(but I'm not sure if we can now tell the errors apart, so likely not feasible anyway)

Reply:
I improved this logic by verifying it calling `file_exists(path)` first.

Reply:
This fells a little race-y, as the directory can be removed between the check and read operation.
But it is correct here, as the fact of directory not existing at any point during this call, is a good enough reason for us to return empty vector.

Still, I'd leave a comment that once we have a better mechanism for recognizing Language Server errors one from another, we should use this capability here.